### PR TITLE
feat: Add RenVM BTC Bridge Adapter

### PR DESCRIPTION
**Pull request by DeFiFrog**
Ren adapter built by Badger RenVM integration team. Contains mint/burn logic for renBTC and wBTC.

Ren team is ready to launch marketing materials as soon as Thurday. Digg/Shakeshack, let's coordinate when we can push this forward.

### CODE REVIEW

File: contracts/badger-ren/BadgerRenAdapter.sol [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        bytes32 _nHash,
+        bytes calldata _sig
+    ) external nonReentrant {
+        // Mint renBTC tokens
+        bytes32 pHash = keccak256(abi.encode(_slippage, _destination));
+        uint256 _mintAmount = registry.getGatewayBySymbol("BTC").mint(pHash, _amount, _nHash, _sig);
+        _swapRenBTCForWBTC(_slippage, _destination, _mintAmount);
+    }
+
+    function _swapRenBTCForWBTC(
```
Comment (ArcSin2000X) [inline contracts/badger-ren/BadgerRenAdapter.sol:180]: Should author consider implementing swap token method such that it has dependency on smart contract that does the swap therefore allowing for greater modularity and interchangeability in the future? ATM swap is hardcoded to curve.fi which might not be the most optimal place to swap pegged btc around(specially for big amounts). I suspect that in rare but possible scenario where curve pool becomes unbalanced and the user insists on swap with slip limit it will render mint functionality  for wBTC unusable

1inch has off chain facility to find optimal routing for swaps, I am not certain how big of priority is supporting large swaps with limited slip?

File: contracts/badger-ren/BadgerRenAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        uint256 _slippage,
+        address payable _destination,
+        uint256 _mintAmount
+    ) internal {
+        // Get price
+        uint256 dy = exchange.get_dy(0, 1, _mintAmount);
```
Comment (ArcSin2000X) [inline contracts/badger-ren/BadgerRenAdapter.sol:186]: Given we rely 100% on curve.fi to give us the best price(`get_dy`), if pool is unbalanced or swap amount is large it can lead to unfavourable x rate.  To combat this should we consider interrogating oracle for the best price or alternatively consider https://curve.readthedocs.io/registry-exchanges.html#Swaps.get_best_rate? 

Should we consider augmenting user experience with displaying estimated value for the mint operation that involves the swap?

File: contracts/badger-ren/BadgerRenAdapter.sol [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+        _slippage = uint256(1e4).sub(_slippage);
```
Comment (ArcSin2000X) [inline contracts/badger-ren/BadgerRenAdapter.sol:187]: consider asserting `_slippage` value is within its boundary condition to prevent over and under flows as well as unfavourable swaps

File: contracts/badger-ren/BadgerRenAdapter.sol [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+        uint256 _startBalance = wBTC.balanceOf(address(this));
+        try exchange.exchange(0, 1, _mintAmount, min_dy)  {
+            uint256 _endBalance = wBTC.balanceOf(address(this));
+            uint256 _wbtcAmount = _endBalance.sub(_startBalance);
+            uint256 _fee = _processFee(wBTC, _wbtcAmount, mintFeeBps);
+            emit MintWBTC(_mintAmount, _wbtcAmount, _fee);
+
+            // Send converted wBTC to user.
+            require(wBTC.transfer(_destination, _wbtcAmount.sub(_fee)));
```
Comment (ArcSin2000X) [inline contracts/badger-ren/BadgerRenAdapter.sol:198]: inconsistent use of native transfer whereby `_processFee(...)` method is implementing `_safeTransfer(...)`

File: contracts/badger-ren/BadgerRenAdapter.sol [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    }
+
+    function burnWBTC(
+        bytes calldata _btcDestination,
+        uint256 _amount,
+        uint256 _minAmount
+    ) external nonReentrant {
+        wBTC.safeTransferFrom(msg.sender, address(this), _amount);
+        uint256 _startBalance = renBTC.balanceOf(address(this));
+        exchange.exchange(1, 0, _amount, _minAmount);
```
Comment (ArcSin2000X) [inline contracts/badger-ren/BadgerRenAdapter.sol:214]: Here I think we should use the common swap method as it feels risky calling `exchange` method outside of try/catch and without assertion making sure user is not slipping on the swap. I appreciate there is _minAmount passed as argument however I think we could refactor this method to be modular like `_swapRenBTCForWBTC`.

File: contracts/badger-ren/BadgerRenAdapter.sol [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+            emit ExchangeWBTCBytesError(_error);
+        }
+        _fallbackTransferRenBTC(_destination, _mintAmount);
+    }
+
+    function burnWBTC(
+        bytes calldata _btcDestination,
+        uint256 _amount,
+        uint256 _minAmount
+    ) external nonReentrant {
```
Comment (ArcSin2000X) [inline contracts/badger-ren/BadgerRenAdapter.sol:211]: Consider asserting `_minAmount` is not going to cause too much damage to users swap, I appreciate this might be ambiguous ask but we should try and pretend user value over the swap

File: tests/ren-integration/adapter_test.js [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+        // is not estimatable.
+        txConfig: { gas: 1000000 },
+      })),
+    };
+
+    const burn = await renJS.burnAndRelease(params);
+    logger.info('processing renBTC burn...');
+    await processBurn(burn);
+  });
+
```
Comment (ArcSin2000X) [inline tests/ren-integration/adapter_test.js:172]: Should we consider adding scenarios:
1. Pool unbalanced
2. Negative test cases: arguments outside of boundary conditions to test assertions prevent errors

File: tests/ren-integration/adapter_test.js [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+        // NB: Need to include gas for burn tx as the gas cost of the burn method
+        // is not estimatable.
+        txConfig: { gas: 1000000 },
+      })),
+    };
+
+    const burn = await renJS.burnAndRelease(params);
+    logger.info('processing renBTC burn...');
+    await processBurn(burn);
+  });
```
Comment (ArcSin2000X) [inline tests/ren-integration/adapter_test.js:171]: Consider assertion rules to test for desired results

File: contracts/badger-ren/BadgerRenAdapter.sol [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+    uint256 public constant MAX_BPS = 10000;
+
+    function initialize(
+        address _governance,
+        address _rewards,
+        address _registry,
+        address _exchange,
+        address _wbtc,
+        uint256[4] memory _feeConfig
+    ) public {
```
Comment (ArcSin2000X) [inline contracts/badger-ren/BadgerRenAdapter.sol:96]: Minor risk: suggestion to check the values on ctor do not have null addresses or values outside of boundary condition

**Review by ArcSin2000X — COMMENTED** [review]
In essence I think we should consider improving func around swaps, but that decision can be made based on priority of my findings. I think for tests we could benefit from some negative scenarios as well as assertion rules

**Comment by notbdu** [comment]
Changing to WIP pending refactor.

**Comment by notbdu** [comment]
Needs some updated e2e integration tests. UI code will need to be updated as well.

- test swap routing/swapping w/ curve strategy (will need to mock caller/ren contracts on mainnet)
- test mint/burn renBTC on kovan (should be less work, just modifying the existing js e2e integration tests)

File: contracts/badger-ren/strategies/CurveSwapStrategy.sol [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+    function swapTokens(
+        address _from,
+        address _to,
+        uint256 _amount,
+        uint256 _slippage
+    ) external nonReentrant onlySwapper returns (uint256 amount) {
+        (address registry, address pool, uint256 exchangeAmount) = _estimateSwapAmount(_from, _to, _amount);
+
+        _slippage = uint256(1e4).sub(_slippage);
```
Comment (ArcSin2000X) [inline contracts/badger-ren/strategies/CurveSwapStrategy.sol:43]: line 43 and 44 are same exact logic as seen in `_minAmount` on `BadgerRenAdapter`, perhaps consider moving it to common method?

File: contracts/badger-ren/strategies/CurveSwapStrategy.sol [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+import "deps/@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import "deps/@openzeppelin/contracts-upgradeable/utils/ReentrancyGuardUpgradeable.sol";
+import "deps/@openzeppelin/contracts-upgradeable/math/SafeMathUpgradeable.sol";
+import "deps/@openzeppelin/contracts-upgradeable/token/ERC20/SafeERC20Upgradeable.sol";
+import "deps/@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+
+import "interfaces/curve/ICurveExchange.sol";
+
+contract CurveSwapStrategy is AccessControlUpgradeable, ReentrancyGuardUpgradeable {
```
Comment (ArcSin2000X) [inline contracts/badger-ren/strategies/CurveSwapStrategy.sol:13]: Should you consider creating interface for all future swap strategies to follow coherent pattern?

File: contracts/badger-ren/strategies/CurveSwapStrategy.sol [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    function _approveBalance(
+        address _token,
+        address _spender,
+        uint256 _amount
+    ) internal {
+        if (IERC20Upgradeable(_token).allowance(address(this), _spender) < _amount) {
+            // Approve max spend.
+            IERC20Upgradeable(_token).approve(_spender, (1 << 64) - 1);
```
Comment (ArcSin2000X) [inline contracts/badger-ren/strategies/CurveSwapStrategy.sol:98]: what is the rationale for approving max spend?

**Review by ArcSin2000X — COMMENTED** [review]
Given the gravity of changes I think we should focus testing on:
1. Atomicity of transactions involving failed burn/mint/swap
2. Edge cases on on slipping pools

File: contracts/badger-ren/BadgerRenAdapter.sol [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+        // darknode args
+        uint256 _amount,
+        bytes32 _nHash,
+        bytes calldata _sig
+    ) external nonReentrant {
+        require(_token == address(renBTC) || _token == address(wBTC), "invalid token address");
+
+        // Mint renBTC tokens
+        bytes32 pHash = keccak256(abi.encode(_token, _slippage, _destination));
+        uint256 mintAmount = registry.getGatewayBySymbol("BTC").mint(pHash, _amount, _nHash, _sig);
```
[the remark this answers is not in the document]
Comment (notbdu) [reply contracts/badger-ren/BadgerRenAdapter.sol:110]: I think the actual mint amount will be less as renVM takes fees IIRC.

File: contracts/badger-ren/BadgerRenAdapter.sol [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+        uint256 fee = _processFee(renBTC, mintAmount, mintFeeBps);
+
+        uint256 wbtcExchanged;
+        if (_token == address(wBTC)) {
+            // Try and swap and transfer wbtc if token ID wbtc specified.
+            uint256 startBalance = wBTC.balanceOf(address(this));
+            if (_swapRenBTCForWBTC(mintAmount.sub(fee), _slippage)) {
```
[the remark this answers is not in the document]
Comment (notbdu) [reply contracts/badger-ren/BadgerRenAdapter.sol:117]: Fee is calculating as a percentage of amount (in basis points) so it should never be more than mint amount. There is also a minimum required mint amount in renJS as well.

Also, doesn't safemath protect against over/underflows (should revert)?

File: contracts/badger-ren/BadgerRenAdapter.sol [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+                uint256 endBalance = wBTC.balanceOf(address(this));
+                wbtcExchanged = endBalance.sub(startBalance);
+                wBTC.safeTransfer(_destination, wbtcExchanged);
+                return;
```
[the remark this answers is not in the document]
Comment (notbdu) [reply contracts/badger-ren/BadgerRenAdapter.sol:121]: Ah yea, we need two mint events.

File: contracts/badger-ren/BadgerRenAdapter.sol [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+
+        if (_token == address(renBTC)) {
+            require(renBTC.balanceOf(address(msg.sender)) >= _amount);
+            renBTC.safeTransferFrom(msg.sender, address(this), _amount);
+        }
+
+        if (_token == address(wBTC)) {
+            wBTC.safeTransferFrom(msg.sender, address(this), _amount);
+            wbtcTransferred = _amount;
+            _swapWBTCForRenBTC(_amount, _slippage);
```
[the remark this answers is not in the document]
Comment (notbdu) [reply contracts/badger-ren/BadgerRenAdapter.sol:150]: There is no return value, this is expected to revert if it fails.

File: contracts/badger-ren/BadgerRenAdapter.sol [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+        }
+
+        uint256 endBalance = renBTC.balanceOf(address(this));
+        uint256 toBurnAmount = endBalance.sub(startBalance);
+        uint256 fee = _processFee(renBTC, toBurnAmount, burnFeeBps);
+
+        emit Burn(toBurnAmount, wbtcTransferred, fee);
```
[the remark this answers is not in the document]
Comment (notbdu) [reply contracts/badger-ren/BadgerRenAdapter.sol:157]: I don't think it matters, events are reverted if the tx fails?

File: contracts/badger-ren/strategies/CurveSwapStrategy.sol (hunk 12 above)
Comment (notbdu) [reply contracts/badger-ren/strategies/CurveSwapStrategy.sol:98]: Save on gas?

File: contracts/badger-ren/strategies/CurveSwapStrategy.sol (hunk 11 above)
Comment (notbdu) [reply contracts/badger-ren/strategies/CurveSwapStrategy.sol:13]: Ah yea I have an `ISwapStrategy` already, I can make it inherit from it.

File: contracts/badger-ren/BadgerRenAdapter.sol (hunk 16 above)
[the remark this answers is not in the document]
Comment (ArcSin2000X) [reply contracts/badger-ren/BadgerRenAdapter.sol:150]: ah, my bad, looked at the other swap method

File: contracts/badger-ren/BadgerRenAdapter.sol (hunk 14 above)
[the remark this answers is not in the document]
Comment (ArcSin2000X) [reply contracts/badger-ren/BadgerRenAdapter.sol]: ok

**Review by dapp-whisperer — APPROVED** [review]
